draft/XXXX-upload-system/README.md

```diff
@@ -0,0 +1,126 @@
+
+```
+POST /api/upload/csv/{workspace}/{table}
+POST /api/upload/d3_json/{workspace}/{graph}
+POST /api/upload/nested_json/{workspace}/{graph}
+POST /api/upload/newick/{workspace}/{graph}
+```
+
+These endpoints replaces the original upload endpoints for each data type we currently support. This returns a JSON response, containing the upload information, in the format referenced above.
+
+A new endpoint for uploading data chunks themselves will have the following format (endpoint for each upload format specified):
+
+```
+PUT /api/upload/csv/{workspace}/{table}
+PUT /api/upload/d3_json/{workspace}/{graph}
+PUT /api/upload/nested_json/{workspace}/{graph}
+PUT /api/upload/newick/{workspace}/{graph}
+
+Body: A chunk of the original data
+
+Query Parameters: {
+  upload: id (returned by the first call),
+}
+```
+
+The `upload` query parameter is part of the information returned by the first call, and is used to denote the upload object to which the sent chunk of data belongs to. Since an upload is tied to a specific user, this endpoint is gated behind user login. Furthermore, a `PUT` request (by a logged in user) to an upload endpoint targeting an upload that does not belong to this user, should return `403 Forbidden`.
+
+
+New endpoints for getting upload information are also required. These would be of the following formats:
+
+```
+GET /api/upload/uploads
+```
+
+This endpoint is tied to the currently authenticated user, and returns all of the uploads which belong to the user. If the need for a global upload listing endpoint arises, this may have to be revised.
+
+
+### Intermediate Processing of Chunks
+
+Each data type uploader (csv, newick, etc.) will be responsible for specific logic regarding how chunked uploads are handled and stored in the "hidden" arangodb collection. Given a chunk of data, the uploader needs to be able to process/keep track the following information:
+
+* How many rows/documents are fully contained in that chunk
+* How to convert/transform a row/document from its original format, into the format required for arangodb storage
+* The current position or "context" in the full document (in graphs for example, are we currently looking at nodes?, links?, etc.)
+
+More than likely, a chunk will not often start/end exactly on the boundary between rows/documents. In this case, any "incomplete" rows/documents will need to stored in a specific location (perhaps on the upload document) until the data which completes that record arrives. Storing this data in memory is not an option, as compatibility across multiple workers must be ensured.
+
+The approach above assumes that these chunks of data are being sent **in order**. If there is a need for sending these chunks out of order, more logic would be required to keep track of "incomplete" rows/documents, and ensuring that in the future, they are able to be completed properly.
```

Comment:
What would be a use case that demands out-of-order chunking? I would be ok with enforcing in-order, solely to avoid this extra logic.

Reply:
Maybe parallel POSTs to speed up the upload of larger files?

Reply:
See [this comment](https://github.com/multinet-app/multinet-rfcs/pull/7#discussion_r452413501), which explains how you can do parallel chunk uploads for efficiency.